### DISCUSSION TITLE
SV06 SKR mini E3 V3 runout sensor

### DISCUSSION THREAD

**Discussion by Astrolava** [bassamanator/Sovol-SV06-firmware#170] [General]
I hooked up a filament runout sensor to the toolhead of my SV06, which uses an SKR Mini E3 V3, I'm very sure it's wired correctly.
My printer.cfg has the following: 

[filament_switch_sensor filament_sensor]
switch_pin: !PA4 # "Pulled-high"
pause_on_runout: True
insert_gcode:
    M117 Insert Detected
    { action_respond_info("Insert Detected") }
runout_gcode:
    M117 Runout Detected
    { action_respond_info("Runout Detected") }
    CONDITIONAL_BEEP i=3 dur=300 freq=400
    
    I can see there's a "filament sensor" part on Klipper's dashboard now, but it says "empty" no matter if there's filament loaded or not.
    I don't seem to have any indication of a state change at all, and I'm wondering if I missed a step? 
    It's a brand new filament runout sensor, but I don't even know how to test it to make sure the actual piece of hardware works.
    
The filament runout sensor is a cheap one off alieepress, so it's not at all impossible I got a faulty one.
I've tested the wire to confirm all 3 connections work.  

Just to clarify the wiring.  On the toolhead, V is top, G is middle, S is bottom.
On the runout sensor, input side is V, middle is G, output side is S.

Any advice on how I might get this working?

**Comment by jakebullet70** [comment]
I just setup one on an SKR 1.4 board on another printer. I used these docs.

https://www.klipper3d.org/G-Codes.html#filament_switch_sensor

I used 'QUERY_FILAMENT_SENSOR' command from the terminal until I had it working. For me it was the wrong pin.

**Comment by Astrolava** [comment]
> I just setup one on an SKR 1.4 board on another printer. I used these docs.
> 
> https://www.klipper3d.org/G-Codes.html#filament_switch_sensor
> 
> I used 'QUERY_FILAMENT_SENSOR' command from the terminal until I had it working. For me it was the wrong pin.

What pin did it end up being for you?
I have just finished testing pins PA-PF (0-15) - aside from the reserved and already used pins.
None of the pins detected the filament runout button being toggled.

I wonder if maybe there's an incorrect pin set somewhere which is occupying the pin for the sensor.  
I noticed my screen's button doesn't work on PA15, but, it also didn't detect the filament runout.
I'm not too sure what to do now.

**Comment by Astrolava** [comment]
These are the already occupied pins which I can't test for obvious reasons.  I wonder if one of these might be the actual probe pin, and if so, what that pin might otherwise be.  
I'm interested in this PA0...

PA0 - [extruder] - sensor_pin
PA9 - [display] - encoder_pins
PA10 - [display] - encoder_pins
PA11 - reserved for USB
PA12 - reserved for USB
PA15 - [display] - click_pin (tested)

PB0 - [stepper_z] - step_pin
PB1 - [stepper_z] - enable_pin
PB2 - [stepper_y] - dir_pin
PB3 - [extruder] - step_pin
PB4 - [extruder] - dir_pin
PB5 - [output_pin beeper] pin
PB8 - [display] - cs_pin
PB9 - [display] - sclk_pin
PB10 - [stepper_y] - step_pin
PB11 - [stepper_y] - enable_pin
PB12 - [stepper_x] - dir_pin
PB13 - [stepper_x] - step_pin
PB14 - [stepper_x] - enable_pin
PB15 - [controller_fan electronic_enclosure_fan] - pin

PC0 - [tmc2209 stepper_x] - diag_pin
PC1 - [tmc2209 stepper_y] - diag_pin
PC4 - [heater_bed] - sensor_pin
PC5 - [stepper_z] - dir_pin
PC6 - [fan] pin
PC7 - [heater_fan hotend_fan] - pin
PC8 - [extruder] - heater_pin
PC9 - [heater_bed] - heater_pin
PC10 - [tmc2209 extruder] tx_pin
PC11 - [tmc2209 extruder] uart_pin
PC14 - [probe] - pin

PD1 - [extruder] - enable_pin
PD6 - [display] - sid_pin

PF0 - reserved for crystal?
PF1 - reserved for crystal?

**Comment by Astrolava** [comment]
I guess another option here might be that the place where the P7 is plugged in requires more configuration.
If the port on the board, which I believe is labeled "DET (power detection?)" detects power being input, then it might require a source of power (3v or so?)  - but, if the DET doesn't also supply that power, then the filament sensor - which is just a button that joins two wires essentially - is just a dead wire.
Doesn't the output (DET) need to supply 3v to Vcc, then detect 3v in ground?  Is that how it works?

Right now, it seems that it's just a dead wire.  I couldn't get any reading on any of the pins of P7.

**Comment by bassamanator** [comment]
You need to uncomment the filament sensor section in `printer.cfg`, as well as adjusting `misc-macros.cfg`. Please see [What are some settings that I can change?](https://github.com/bassamanator/Sovol-SV06-firmware/blob/skr-mini-e3-v3/README.md#what-are-some-settings-that-i-can-change). Read full docs before proceeding.

Ignore what it says on the dashboard. Start a print with no filament in the nozzle, but a piece of filament in the sensor. Remove the filament from the sensor to see if things are working as expected.

**Comment by Astrolava** [comment]
> You need to uncomment the filament sensor section in `printer.cfg`, as well as adjusting `misc-macros.cfg`. Please see [What are some settings that I can change?](https://github.com/bassamanator/Sovol-SV06-firmware/blob/skr-mini-e3-v3/README.md#what-are-some-settings-that-i-can-change). Read full docs before proceeding.
> 
> Ignore what it says on the dashboard. Start a print with no filament in the nozzle, but a piece of filament in the sensor. Remove the filament from the sensor to see if things are working as expected.

Am I still doing the pin PA4 ?

I have this: 

[filament_switch_sensor filament_sensor]
switch_pin: !PA4 # "Pulled-high"
pause_on_runout: True
insert_gcode:
    M117 Insert Detected
runout_gcode:
    M117 Runout Detected
    CONDITIONAL_BEEP i=3 dur=300 freq=400
    
Do you know specifically what needs to be adjusted in misc-macros?  I mean, I was doing the gcode:
QUERY_FILAMENT_SENSOR SENSOR=filament_sensor,
But, it wouldn't change when I'd insert / remove filament.

I did wire the SKR mini E3 V3 based on the diagram you posted.
I'll give it another try while starting a print.  I'll then see if it stops the print.  It didn't originally, which is what made me start testing other things to begin with.  I'll let you know if I make progress.

**Comment by Astrolava** [comment]
So yeah, after testing and testing, it doesn't stop the print.  Doesn't pause it, doesn't beep, no indication at all that anything has changed.
It'll just keep printing without filament.  
I will say this, I attached the filament sensor and wire to a breadboard and was able to toggle an LED using the switch.  The switch is working, the wires are fine.  There's something else wrong here.

**Comment by jakebullet70** [comment]
Had the same issue. You need to turn it on in the START gcode.

**Comment by Astrolava** [comment]
> Had the same issue. You need to turn it on in the START gcode.

Thanks for the reply.  Umm, sorry I actually don't know how to turn it on.  Mind elaborating? 
I'm so confused.  Am I still using PA4?  What's ^PA4?  Is that what I need to do?  
Do you mean to put this in my start gcode? 
SET_FILAMENT_SENSOR SENSOR=filament_sensor ENABLE=1

**Comment by jakebullet70** [comment]
Yes that is right: SET_FILAMENT_SENSOR SENSOR=filament_sensor ENABLE=1

When setting up a filament sensor you do not need to print, you can use 'QUERY_FILAMENT_SENSOR' command and put filament in the sensor by hand, that tells you if you have the correct pin setup, once that is done and QUERY_FILAMENT_SENSOR works by telling when it has filament or not then you turn it on in your START gcode.

**Comment by bassamanator** [comment]
> Am I still doing the pin PA4 ?

@Astrolava revert to whatever was already in `printer.cfg`.

Do not add `SET_FILAMENT_SENSOR SENSOR=filament_sensor ENABLE=1` but if you think you know better than the author of this config, go for it.

**Comment by jakebullet70** [comment]
I sad that is what  had to do on an SKR 1.4 board. Did not work with out it.  Sorry,  I wll back out.

**Comment by bassamanator** [comment]
@jakebullet70 No need to back out. The issue is that you'll have a hard time keeping up with any changes in this config. The only thing my users are required to do are to either uncomment/comment out a section, or change a 1 to a 0 or vice versa. No programming knowledge is *required*.

**Comment by Astrolava** [comment]
I'm... not sure why it doesn't work.  I'll check to see if my ribbon cable is maybe busted.
I've still got the P7 thingy hooked up to I believe DET on the SKR mini E3 V3, which someone said was PC12.  But, yeah, PA4, PC12, doesn't matter, nothing works.  I'll just have to check the ribbon cable.

**Comment by bassamanator** [comment]
Updated wiring diagram, printer.cfg and misc-macros.cfg. Pull new changes.

1. Take note of the wiring connecting to the SKR's `e0-stop`.
2. Take note of how the filament sensor is connected to the hotend pcb.
3. Take note here: https://github.com/bassamanator/everything-sovol-sv06?tab=readme-ov-file#filament-sensor
4. Enable filament sensor via: https://github.com/bassamanator/Sovol-SV06-firmware/blob/skr-mini-e3-v3/README.md#what-are-some-settings-that-i-can-change
5. Restart firmware and query the sensor via @jakebullet70's suggestion, `QUERY_FILAMENT_SENSOR SENSOR=filament_sensor`. 

Can't simplify the process any further.

**Comment by PhoenixWolf27** [comment]
May not be relevant to @Astrolava's issue if it's more hardware than software, but I was troubleshooting my runout sensor that used to work and then suddenly stopped working. 

TL;DR - make sure you're putting "PRINT_START" in the start g-code in your slicer of choice. There's a macro command in there called ADJUST_FILAMENT_SENSOR_STATUS ENABLE=1 that needs to run otherwise it seems the default state of the sensor is turned off despite turning it on via "variable_filament_sensor_enabled: 1" in @bassamanator's misc-macros.cfg. At least for me. 

I started using Christian Vick's "[Klipper Printer Additions](https://www.printables.com/model/585691-klipper-printer-additions-30b91-more-than-boring-s)", which adds some nice macros and other elements. But his OrcaSlicer (and other slicer) profiles are alternates to Bassamanator's "PRINT_START". 

It took me a while to figure out what was missing, but ultimately rather than try to align the different profiles start commands, I just added ADJUST_FILAMENT_SENSOR_STATUS ENABLE=1 directly to my start g-code, and that seems to have done the trick for me. More of an issue on my end for mucking about with settings.

**Comment by Astrolava** [comment]
I've been dealing with trying to get my SKR mini E3 V3 working with the uart pins of the Orange Pi 3 LTS (running mainsailOS(armbian)).
I finally got it working, but alas, even after reinstalling klipper, mainsail, all that - and reflashing the SKR mini E3 V3 board, I connected the DET cable to the E0-STOP pins.  
But, the Filament runout sensor doesn't trigger any change when queried, nor on firmware restart.
There's a blue LED on my filament runout sensor which is clearly on when there's no filament, and off when there is filament.
I have it wired according to the P7 part of the extruder board.
I'm sure it's wired correctly, or else the LED wouldn't light up.

my printer.cfg:
[filament_switch_sensor filament_sensor]
switch_pin: PC15 # NOTE Add a ! to the pin to reverse if you're getting the opposite behaviour; !PC15
pause_on_runout: True
insert_gcode:
    { action_respond_info("Insert Detected") }
runout_gcode:
    { action_respond_info("Runout Detected") }
    CONDITIONAL_BEEP i=3 dur=300 freq=400
    

Under misc-macros.cfg I changed the delayed_gcode to enable the filament sensor rather than disable it, 
though the only thing it changed was it caused the filament runout sensor to be enabled after restarting the firmware
It had no impact on whether or not the runout sensor could trigger:

[delayed_gcode DISABLEFILAMENTSENSOR] 
initial_duration: 1
gcode:
    SET_FILAMENT_SENSOR SENSOR=filament_sensor ENABLE=1

[gcode_macro _globals]
variable_filament_sensor_enabled: 1 # NOTE Enable(1) or disable(0) the filament sensor, if one is connected
variable_beeping_enabled: 1 # NOTE Enable(1) or disable(0) beeping everywhere except during gantry calibration
variable_bed_temp_over: 10 # NOTE Start print if bed temperature is over by this amount, otherwise wait for temperature drop
variable_kamp_enable: 1 # NOTE Enable(1) or disable(0) KAMP (adaptive mesh)
variable_pre_purge_prime_length: 1.40
gcode:

I've included some pictures which show the wiring, and the sensor.  
If anyone can see any mistakes, or if maybe there's some type of configuration I need to enable on the orange pi 3 LTS, please let me know.  I'm kind of at a loss as to what to do.
Just another note - for the DET cable that plugs into the board, I did extend the length of the wires, but I did not re-crimp the connector, as well, since I extended each wire one at a time, there is no chance these wires got crossed during the process.
Maybe some pins need to be enabled?  Or disabled?  I have no idea.
Maybe my Orange pi 3 LTS is somehow not responding to the signal from the SKR mini E3 V3 on the E0-STOP?
One more note.  I tried !P15, ^P15, P15, even tried P12, ^P12, !P12, but it made absolutely no difference.
Please help!




![20240521_200708](https://github.com/bassamanator/Sovol-SV06-firmware/assets/45184363/45df5f6b-5268-4fcc-afee-5f38b24477d1)
![20240521_200717](https://github.com/bassamanator/Sovol-SV06-firmware/assets/45184363/83b27162-4a3f-48e4-beb4-37abc0fc6b0a)
![20240521_200817](https://github.com/bassamanator/Sovol-SV06-firmware/assets/45184363/0b1b3a25-9666-4732-a224-a8b922b7b1a4)

**Comment by bassamanator** [comment]
https://www.reddit.com/r/Sovol/comments/1631jk5/comment/l51zza4/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

Here's the situation:
1. You've made changes to the config that broke it.
2. You didn't wire things correctly.
3. The filament sensor is at fault.

Have your pick. And you know, these sensors are cheap, $1.5, I had 2 lying around that just worked (after I wired them correctly), but that doesn't mean all will work. Buy a *different* $1.5 sensor and give that a go.
